docs: add basic info on using playwright for plugin development (#32094)
backstage/backstage: docs/plugins/playwright.md

---
id: playwright
title: Testing with Playwright
description: Testing frontend plugins with Playwright
---

Testing frontend plugins UI can be done using [Playwright](https://playwright.dev/).

> Playwright is a UI test framework built by Microsoft, and comes in with support for

Reviewer: ```suggestion
Playwright is a UI test framework built by Microsoft, and out of the box comes with support for
```

all major browsers and platforms.

## Installing Playwright

Add the playwright test library to the workspace:

>     yarn add -D @playwright/test

Reviewer: ```suggestion
```bash
yarn add -D @playwright/test
\```
```

Install runners and browsers (e.g. Chrome)

>     yarn playwright install --with-deps chrome

Reviewer: ```suggestion
```bash
yarn playwright install --with-deps chrome
\```
```

  retries: process.env.CI ? 2 : 0,

  reporter: [['html', { open: 'never', outputFolder: 'e2e-test-report' }]],

  // point to 'http://localhost:3000' by default
  use: {
    baseURL: process.env.PLAYWRIGHT_URL ?? 'http://localhost:3000',
    screenshot: 'only-on-failure',
    trace: 'retain-on-failure',
  },

  // test on chrome
  projects: [
    {
      use: {
        channel: 'chrome',
      },
    },
  ],
});
```

For full list of available options, see the [documentation](https://playwright.dev/docs/test-configuration).

## Writing test cases

Playwright test cases consist of taking actions, and performing assertions.

Example test case:

```typescript
import { test, expect } from '@playwright/test';

test('Guest login to backstage works', async ({ page }) => {
  await page.goto('http://localhost:3000');
  await page.getByRole('button', { name: 'Enter' }).click();

  await expect(page.getByRole('heading', { name: 'Catalog' })).toBeVisible();
});
```

### Locating elements

The `page` fixture API offers multiple built-in [locators](https://playwright.dev/docs/locators#quick-guide).
These include locating by text, role, label, or other user-facing attributes.

> While locating by attributes such as CSS or Xpath is available, it is discouraged in favour of the user-facing locators.

Reviewer: ```suggestion
While locating by attributes such as CSS or XPath is available, it is discouraged in favour of the user-facing locators.
```
if vale still isn't happy, you likely have to add this to https://github.com/backstage/backstage/blob/d78945eff946cb8f60e8b7ee92062eadd223ca9e/.github/vale/config/vocabularies/Backstage/accept.txt#L1
